### PR TITLE
setStateBasedEndorsement NOutOf for typescript chaincode

### PR DESCRIPTION
**Pull request by girigaurav**
Signed-off-by: Gaurav Giri <girigaurav@gmail.com>

### CODE REVIEW

**Review by lehors — CHANGES_REQUESTED** [review]
Seems to work fine but I'd like to see a few clarifications and one change to the code.

File: asset-transfer-sbe/chaincode-java/src/main/java/org/hyperledger/fabric/samples/sbe/AssetContract.java [hunk 1]
```diff
@@ -71,10 +71,10 @@ public Asset CreateAsset(final Context ctx, final String assetId, final int valu
         stub.putStringState(assetId, assetJSON);
 
         // Set the endorsement policy of the assetId Key, such that current owner Org Peer is required to endorse future updates
-        setAssetStateBasedEndorsement(ctx, assetId, new String[]{ownerOrg});
+        setStateBasedEndorsement(ctx, assetId, new String[]{ownerOrg});
 
-        // Optionally, set the endorsement policy of the assetId Key, such that any 1(N) out of the Org's specified can endorse future updates
-        // setAssetStateBasedEndorsementWithNOutOfPolicy(ctx, assetId, 1, new String[]{"Org1MSP", "Org2MSP"});
+        // Optionally, set the endorsement policy of the assetId Key, such that any 1(N) out of the Org's Peer specified can endorse future updates
```
[the remark this answers is not in the document]
Comment (girigaurav) [reply asset-transfer-sbe/chaincode-java/src/main/java/org/hyperledger/fabric/samples/sbe/AssetContract.java]: done

File: asset-transfer-sbe/chaincode-java/src/main/java/org/hyperledger/fabric/samples/sbe/AssetContract.java [hunk 2]
```diff
@@ -216,12 +216,12 @@ private static void setAssetStateBasedEndorsement(final Context ctx, final Strin
...
      * @param ownerOrgs the list of Owner Org MSPID's
      */
-    private static void setAssetStateBasedEndorsementWithNOutOfPolicy(final Context ctx, final String assetId, final int nOrgs, final String[] ownerOrgs) {
+    private static void setStateBasedEndorsementNOutOf(final Context ctx, final String assetId, final int nOrgs, final String[] ownerOrgs) {
         ctx.getStub().setStateValidationParameter(assetId, policy(nOrgs, Arrays.asList(ownerOrgs)));
     }
 
     /**
-     * Create the policy such that it requires any N signature's from all of the principals provided
+     * Create the policy such that it requires any N Orgs signature's from all of the principals provided
```
[the remark this answers is not in the document]
Comment (girigaurav) [reply asset-transfer-sbe/chaincode-java/src/main/java/org/hyperledger/fabric/samples/sbe/AssetContract.java]: done

File: asset-transfer-sbe/chaincode-java/src/main/java/org/hyperledger/fabric/samples/sbe/AssetContract.java [hunk 3]
```diff
@@ -216,12 +216,12 @@ private static void setAssetStateBasedEndorsement(final Context ctx, final Strin
...
      *
      * @param nOrgs the number of N Org signature's to endorse out of the list of Orgs provided
```
[the remark this answers is not in the document]
Comment (girigaurav) [reply asset-transfer-sbe/chaincode-java/src/main/java/org/hyperledger/fabric/samples/sbe/AssetContract.java]: done

File: asset-transfer-sbe/chaincode-typescript/src/assetContract.ts [hunk 4]
```diff
@@ -26,8 +27,12 @@ export class AssetContract extends Contract {
         const buffer = Buffer.from(JSON.stringify(asset));
         // Create the asset
         await ctx.stub.putState(assetId, buffer);
+
         // Set the endorsement policy of the assetId Key, such that current owner Org Peer is required to endorse future updates
```
[the remark this answers is not in the document]
Comment (girigaurav) [reply asset-transfer-sbe/chaincode-typescript/src/assetContract.ts]: done

File: asset-transfer-sbe/chaincode-typescript/src/assetContract.ts [hunk 5]
```diff
@@ -26,8 +27,12 @@ export class AssetContract extends Contract {
...
-        await AssetContract.setAssetStateBasedEndorsement(ctx, asset.ID, [ownerOrg]);
+        await AssetContract.setStateBasedEndorsement(ctx, assetId, [ownerOrg]);
+
+        // Optionally, set the endorsement policy of the assetId Key, such that any 1(N) out of the Org's Peer specified can endorse future updates
```
[the remark this answers is not in the document]
Comment (girigaurav) [reply asset-transfer-sbe/chaincode-typescript/src/assetContract.ts]: done

File: asset-transfer-sbe/chaincode-typescript/src/assetContract.ts [hunk 6]
```diff
@@ -87,16 +92,58 @@ export class AssetContract extends Contract {
...
-    // setAssetStateBasedEndorsement sets an endorsement policy to the assetId Key
-    // setAssetStateBasedEndorsement enforces that the owner Org Peers must endorse future update transactions for the specified assetId Key
-    private static async setAssetStateBasedEndorsement(ctx: Context, assetId: string, ownerOrgs: string[]): Promise<void> {
+    // getClientOrgId gets the client's OrgId (MSPID)
+    private static getClientOrgId(ctx: Context): string {
+        return ctx.clientIdentity.getMSPID();
+    }
+
+    // setStateBasedEndorsement sets an endorsement policy to the assetId Key
+    // setStateBasedEndorsement enforces that the owner Org Peers must endorse future update transactions for the specified assetId Key
```
[the remark this answers is not in the document]
Comment (girigaurav) [reply asset-transfer-sbe/chaincode-typescript/src/assetContract.ts]: done

File: asset-transfer-sbe/chaincode-typescript/src/assetContract.ts [hunk 7]
```diff
@@ -87,16 +92,58 @@ export class AssetContract extends Contract {
...
         const ep = new KeyEndorsementPolicy();
         ep.addOrgs('MEMBER', ...ownerOrgs);
         await ctx.stub.setStateValidationParameter(assetId, ep.getPolicy());
     }
 
-    // getClientOrgId gets the client's OrgId (MSPID)
-    private static getClientOrgId(ctx: Context): string {
-        return ctx.clientIdentity.getMSPID();
+    // setStateBasedEndorsementNOutOf sets an endorsement policy to the assetId Key
+    // setStateBasedEndorsementNOutOf enforces that any N out of the Org's Peers specified must endorse future update transactions for the specified assetId Key.
```
[the remark this answers is not in the document]
Comment (girigaurav) [reply asset-transfer-sbe/chaincode-typescript/src/assetContract.ts]: done

File: asset-transfer-sbe/chaincode-typescript/src/assetContract.ts [hunk 8]
```diff
@@ -78,7 +83,7 @@ export class AssetContract extends Contract {
         // Update the asset
         await ctx.stub.putState(assetId, Buffer.from(JSON.stringify(asset)));
         // Re-Set the endorsement policy of the assetId Key, such that a new owner Org Peer is required to endorse future updates
-        await AssetContract.setAssetStateBasedEndorsement(ctx, asset.ID, [newOwnerOrg]);
+        await AssetContract.setStateBasedEndorsement(ctx, asset.ID, [newOwnerOrg]);
```
[the remark this answers is not in the document]
Comment (girigaurav) [reply asset-transfer-sbe/chaincode-typescript/src/assetContract.ts:86]: done

File: asset-transfer-sbe/chaincode-java/src/main/java/org/hyperledger/fabric/samples/sbe/AssetContract.java [hunk 9]
```diff
@@ -164,7 +164,7 @@ public Asset TransferAsset(final Context ctx, final String assetId, final String
         String updatedAssetJSON = genson.serialize(asset);
         stub.putStringState(assetId, updatedAssetJSON);
 
-        setAssetStateBasedEndorsement(ctx, assetId, new String[]{newOwnerOrg});
+        setStateBasedEndorsement(ctx, assetId, new String[]{newOwnerOrg});
```
[the remark this answers is not in the document]
Comment (girigaurav) [reply asset-transfer-sbe/chaincode-java/src/main/java/org/hyperledger/fabric/samples/sbe/AssetContract.java:168]: done

**Review by lehors — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**